src/web/releases.rs

```diff
@@ -215,84 +216,84 @@ fn get_releases_by_owner(
-                    releases.crate_id,
-                    RANK() OVER (PARTITION BY crate_id ORDER BY release_time DESC) as rank
-                FROM releases
-                WHERE releases.rustdoc_status AND NOT releases.yanked
-            ) AS releases
-            WHERE releases.rank = 1
-        ) AS latest_release ON latest_release.crate_id = crates.id
-        INNER JOIN releases ON latest_release.id = releases.id
-        LEFT JOIN github_repos ON releases.github_repo = github_repos.id
-        WHERE
-            ((char_length($1)::float - levenshtein(crates.name, $1)::float) / char_length($1)::float) >= 0.65
-            OR crates.name ILIKE CONCAT('%', $1, '%')
-        GROUP BY crates.id, releases.id, github_repos.stars
-        ORDER BY
-            levenshtein(crates.name, $1) ASC,
-            crates.name ILIKE CONCAT('%', $1, '%'),
-            releases.downloads DESC
-        LIMIT $2 OFFSET $3";
-
-    let rows = conn.query(statement, &[&query, &limit, &offset])?;
 
-    // Each row contains the total number of possible/valid results, just get it once
-    let total_results = rows
-        .get(0)
-        .map(|row| row.get::<_, i64>("total"))
-        .unwrap_or_default();
-    let packages: Vec<Release> = rows
+    use crate::utils::APP_USER_AGENT;
+    use once_cell::sync::Lazy;
+    use reqwest::blocking::Client as HttpClient;
+    use reqwest::header::{HeaderMap, HeaderValue, ACCEPT, USER_AGENT};
+
+    static HTTP_CLIENT: Lazy<HttpClient> = Lazy::new(|| {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+        HttpClient::builder()
+            .default_headers(headers)
+            .build()
+            .unwrap()
+    });
+
+    let url = format!(
+        "https://crates.io/api/v1/crates?page={page}&per_page={limit}&q={query}",
+        page = page,
+        limit = limit,
+        query = query
+    );
+    debug!("fetching search results from {}", url);
+    let releases: CratesIoReleases = HTTP_CLIENT.get(&url).send()?.json()?;
+    let query = conn.prepare(
+        "SELECT github_repos.stars, releases.target_name, releases.rustdoc_status
+        FROM crates INNER JOIN releases ON crates.id = releases.crate_id
+                    LEFT JOIN github_repos ON releases.github_repo = github_repos.id
+        WHERE crates.name = $1 AND releases.version = $2",
+    )?;
+    let crates = releases
+        .crates
         .into_iter()
-        .map(|row| Release {
-            name: row.get("name"),
-            version: row.get("version"),
-            description: row.get("description"),
-            target_name: row.get("target_name"),
-            release_time: row.get("release_time"),
-            rustdoc_status: row.get("rustdoc_status"),
-            stars: row.get::<_, Option<i32>>("github_stars").unwrap_or(0),
+        .flat_map(|krate| {
```

Comment:
(caveat, not a rust pro): 

when we query all the details from our db before we jump into handling the API response, 
wouldn't it be easier to read to 

```rust 
let crates = releases 
    .crates
    .into_iter()
    .filter(|&krate| db_data.contains_key((&krate.name, &krate.max_version))
    .map(|krate| {
        ...
    })
```

instead of this `flat_map`, `Option` and `Result` wrapping/unwrapping?

Reply:
Where does `db_data` come from? Right now it's in the `flat_map` itself (`map` in your snippet). If I move it earlier then I now have to handle the `None` error earlier and it ends up as the same code.

Reply:
`db_data` would come from fetching the data from for all crates in the result into (for example) a  `HashMap`. 
Other way would be to turn it around and create the method result based on the sql result, while enriching it with the information from the search result. 

In the end, you are more experienced with rust, perhaps this here is not an issue from a readability standpoint for more experienced rust developers. 

But this was the only part of the implementation where I didn't directly understand what it does and how (Most of my history in years is in Python and C++).